src/plonk/vanishing/verifier.rs

```diff
@@ -32,55 +33,41 @@ impl<C: CurveAffine> Argument<C> {
 }
 
 impl<C: CurveAffine> Committed<C> {
-    pub(in crate::plonk) fn evaluate<T: TranscriptRead<C>>(
-        self,
-        transcript: &mut T,
-    ) -> Result<Evaluated<C>, Error> {
-        let h_evals = read_n_scalars(transcript, self.h_commitments.len())
-            .map_err(|_| Error::TranscriptError)?;
-
-        Ok(Evaluated {
-            h_commitments: self.h_commitments,
-            h_evals,
-        })
-    }
-}
-
-impl<C: CurveAffine> Evaluated<C> {
     pub(in crate::plonk) fn verify(
-        &self,
+        self,
         expressions: impl Iterator<Item = C::Scalar>,
         y: ChallengeY<C>,
         xn: C::Scalar,
-    ) -> Result<(), Error> {
+    ) -> Result<Evaluated<C>, Error> {
         let expected_h_eval = expressions.fold(C::Scalar::zero(), |h_eval, v| h_eval * &*y + &v);
+        let expected_h_eval = expected_h_eval * ((xn - C::Scalar::one()).invert().unwrap());
```

Comment:
It will happen only if x is a 2^n root of unity; this is part of the "disallowed" challenge space (see also #151,  #152, where we need to clear up what's going on with the challenges) but in practice it will be unlikely to occur. There is also another reason `x` is prohibited from being a 2^n root of unity: if the prover supplied `a(x)` in such a scenario for their secret polynomial `a(X)` then it would reveal one of the cell values!